tests/send_file.rs

```diff
@@ -0,0 +1,89 @@
+use once_cell::sync::Lazy;
+use termchat::application::{Application, Event, Config};
+use message_io::events::EventSender;
+use std::io::stdout;
+use crossterm::execute;
+use crossterm::terminal::{EnterAlternateScreen, LeaveAlternateScreen};
+use std::io::Write;
+
+static CONFIG: Lazy<Config> = Lazy::new(|| Config {
+    discovery_addr: "238.255.0.1:5877".parse().unwrap(),
+    tcp_server_port: "0".parse().unwrap(),
+    user_name: "A".to_string(),
+});
+static CONFIG2: Lazy<Config> = Lazy::new(|| Config {
+    discovery_addr: "238.255.0.1:5877".parse().unwrap(),
+    tcp_server_port: "0".parse().unwrap(),
+    user_name: "B".to_string(),
+});
+
+#[test]
+fn send_file() {
+    let _g = Guard::new();
+
+    let termchat_dir = std::env::temp_dir().join("termchat");
+    let test_path = termchat_dir.join("test");
+    let _ = std::fs::remove_dir_all(&termchat_dir);
+    std::fs::create_dir_all(&termchat_dir).unwrap();
+
+    let data = vec![rand::random(); 10usize.pow(6)];
+    std::fs::write(&test_path, &data).unwrap();
+
+    let (mut s1, t1) = test_user(1);
+    let (s2, t2) = test_user(2);
+
+    // wait for users to connect
+    std::thread::sleep(std::time::Duration::from_millis(100));
+    // send file
+    input(&mut s1, &format!("?send {}", test_path.display()));
+    // wait for the file to finish sending
+    std::thread::sleep(std::time::Duration::from_secs(2));
+
+    // finish
+    s1.send(Event::Close(None));
+    s2.send(Event::Close(None));
+    t1.join().unwrap();
+    t2.join().unwrap();
+
+    // assert eq
+    let send_data =
+        std::fs::read(std::env::temp_dir().join("termchat").join("A").join("test")).unwrap();
+    assert_eq!(data.len(), send_data.len());
+    assert_eq!(data, send_data);
+}
+
+fn test_user(n: usize) -> (EventSender<Event>, std::thread::JoinHandle<()>) {
+    let config = if n == 1 { &CONFIG } else { &CONFIG2 };
+    let mut app = Application::new(config).unwrap();
+    let sender = app.sender();
+    let t = std::thread::spawn(move || {
```

Comment:
If you move the `let mut app = Application::new(config).unwrap();` and config inside thread scope, it can work with a reference

Reply:
If you do that you cant extract the sender anymore

Reply:
And pass the config as `&'static Config` to `test_user` doesn't works?

Reply:
You can workaround lifetime issue like this 
```rust
fn test_user(n: usize) -> (EventSender<Event>, std::thread::JoinHandle<()>) {
    let (tx, rx) = std::sync::mpsc::channel();
    let t = std::thread::spawn(move || {
        let config: Config = Config {
            discovery_addr: "238.255.0.1:5877".parse().unwrap(),
            tcp_server_port: "0".parse().unwrap(),
            user_name: n.to_string(),
        };

        let mut app = Application::new(&config).unwrap();
        let sender = app.sender();
        tx.send(sender).unwrap();
        app.run().unwrap();
    });
    (rx.recv().unwrap(), t)
}
```
If you prefer that I can merge this with io::sink() idea

Reply:
Cool! I like the idea. I suppose that if the sender is now extracted by the channel, the config can be passed in the `test_user` params

Reply:
It is ok to merge it with the sink() idea :)